File: src/services/docker/gitGuard.ts
import { spawn } from "node:child_process";
import * as fs from "node:fs";
import * as path from "node:path";

export interface GitRunner {
  exists(cwd: string): boolean;
  run(args: string[], cwd: string): Promise<number>;
}

const realGit: GitRunner = {
  exists: (cwd) => fs.existsSync(path.join(cwd, ".git")),
  run: (args, cwd) =>
    new Promise((resolve) => {
      const c = spawn("git", args, { cwd, stdio: "ignore" });
      c.on("close", (code) => resolve(code ?? 1));
      c.on("error", () => resolve(1));
    }),
};

export interface GitStatusReport {
  refusals: string[];
  warnings: string[];
  skipped: boolean;
}

export async function checkEnvFileGitStatus(
  envFiles: string[],
  cwd: string,
  git: GitRunner = realGit,
): Promise<GitStatusReport> {
  if (!git.exists(cwd)) return { refusals: [], warnings: [], skipped: true };
  const refusals: string[] = [];
  const warnings: string[] = [];
  for (const f of envFiles) {
    const tracked = (await git.run(["ls-files", "--error-unmatch", "--", f], cwd)) === 0;
    if (tracked) {
      refusals.push(f);
      continue;
    }
    const ignored = (await git.run(["check-ignore", "-q", "--", f], cwd)) === 0;
    if (!ignored) warnings.push(f);
  }
  return { refusals, warnings, skipped: false };
}

File: python/src/docker_agent/services/docker/git_guard.py
"""Git status guard for env files.

Parity: ``src/services/docker/gitGuard.ts``.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol


class GitRunner(Protocol):
    def exists(self, cwd: str) -> bool: ...

    async def run(self, args: list[str], cwd: str) -> int: ...


class RealGitRunner:
    def exists(self, cwd: str) -> bool:
        return Path(cwd, ".git").exists()

    async def run(self, args: list[str], cwd: str) -> int:
        proc = await asyncio.create_subprocess_exec(
            "git",
            *args,
            cwd=cwd,
            stdout=asyncio.subprocess.DEVNULL,
            stderr=asyncio.subprocess.DEVNULL,
        )
        await proc.wait()
        return proc.returncode or 0


@dataclass(frozen=True)
class GitStatusReport:
    refusals: list[str]
    warnings: list[str]
    skipped: bool


async def check_env_file_git_status(
    env_files: list[str],
    cwd: str,
    *,
    git: GitRunner | None = None,
) -> GitStatusReport:
    """Check whether env files are tracked, ignored, or untracked by git."""
    runner = git if git is not None else RealGitRunner()
    if not runner.exists(cwd):
        return GitStatusReport(refusals=[], warnings=[], skipped=True)

    refusals: list[str] = []
    warnings: list[str] = []
    for f in env_files:
        tracked = await runner.run(["ls-files", "--error-unmatch", "--", f], cwd)
        if tracked == 0:
            refusals.append(f)
            continue
        ignored = await runner.run(["check-ignore", "-q", "--", f], cwd)
        if ignored != 0:
            warnings.append(f)
    return GitStatusReport(refusals=refusals, warnings=warnings, skipped=False)


__all__ = [
    "GitRunner",
    "GitStatusReport",
    "RealGitRunner",
    "check_env_file_git_status",
]